### PR TITLE
Add perturb() to allow capturing intermediate gradients

### PR DESCRIPTION
**Pull request by IvyZX**
Add a perturb() function in nn.Module, which could be used to capture the intermediate gradients inside a module's __call__.

See example in the docstring. There will be a future Colab/notebook with better examples.

### CODE REVIEW

**Comment by cgarciae** [comment]
Looks good! Should we add a simple test?

**Comment by IvyZX** [comment]
> Looks good! Should we add a simple test?

Test added!

File: flax/linen/module.py [hunk 1]
```diff
@@ -1462,6 +1462,48 @@ def __call__(self, x):
     self.scope.put_variable(col, name, xs)
     return True
 
+  def perturb(self, col: str, name: str, value: T) -> T:
```
Comment (levskaya) [inline flax/linen/module.py:1465]: I wonder if we should encourage a default (but still configurable) collection name "perturbations" or some such here...

**Review by levskaya — APPROVED** [review]
Looks good as a start.  I still think we should think about whether there are ways to simplify using "perturb" in practice w.o. incurring too much extra memory burden.  We may wish to reserve the right to modify this API as we understand how people will use it - perhaps it's worth marking as "experimental" in docstrings to be clear that we may need to tweak it in the future.

File: flax/linen/module.py (hunk 1 above)
Comment (IvyZX) [reply flax/linen/module.py]: Fixed both! Made "collections" a keyword arg with default, and modified the docstring.

**Comment by cgarciae** [comment]
@IvyZX @levskaya  we forgot to add this new method to the sphinx page in `docs/api_reference/flax.linen.rst`, its currently not being rendered.
